SteamKit2/SteamKit2/Util/HardwareUtils.cs

```diff
@@ -281,12 +272,63 @@ public override byte[] GetDiskId()
         }
     }
 
-    static class HardwareUtils
+    /// <summary>
+    /// Class to allow user to provide custom machine info provider.
+    /// </summary>
+    public static class HardwareUtils
     {
+        private static IMachineInfoProvider userMachineInfoProvider;
+        private static bool IsMachineInfoProviderInitialized;
+
+        /// <summary>
+        /// User-provided MachineInfoProvider.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">Occurs when a user tries to (re-)set this property after its first usage.</exception>
+        public static IMachineInfoProvider UserMachineInfoProvider
+        {
+            get => userMachineInfoProvider;
+            set
+            {
+                if ( IsMachineInfoProviderInitialized )
+                {
+                    throw new InvalidOperationException(nameof(UserMachineInfoProvider) + " can't be re-set.");
```

Comment:
This message could confuse a user if they never set it in the first place, but `GetProvider()` was called first. This needs a revised API design.